# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/midi/MIDIPortChild.cpp
# Commit: cfba299e5067
# Full Hash: cfba299e5067a4eab3f3b982374b0770970a3401
# Author: Gabriele Svelto <gsvelto@mozilla.com>
# Date: 2022-04-12 03:57:01
# Regressor Bug: 1758468
# File Overlap Count: 1
# Description:
#   Bug 1758468 - Use stable, anonymous IDs for Web MIDI ports r=padenot
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D142550
# ==============================================================================

diff -r c0eb987d0f76 -r cfba299e5067 dom/midi/MIDIPortChild.cpp
--- a/dom/midi/MIDIPortChild.cpp	Mon Apr 11 17:43:08 2022 +0000
+++ b/dom/midi/MIDIPortChild.cpp	Mon Apr 11 17:43:09 2022 +0000
@@ -7,6 +7,7 @@
 #include "mozilla/dom/MIDIPortChild.h"
 #include "mozilla/dom/MIDIPort.h"
 #include "mozilla/dom/MIDIPortInterface.h"
+#include "nsContentUtils.h"
 
 using namespace mozilla;
 using namespace mozilla::dom;
@@ -55,3 +56,19 @@
   mActorWasAlive = true;
   AddRef();
 }
+
+nsresult MIDIPortChild::GenerateStableId(const nsACString& aOrigin) {
+  const size_t kIdLength = 64;
+  mStableId.SetCapacity(kIdLength);
+  mStableId.Append(Name());
+  mStableId.Append(Manufacturer());
+  mStableId.Append(Version());
+  // Extend to at least 64 characters
+  while (mStableId.Length() < kIdLength) {
+    mStableId.Append(' ');
+  }
+  nsContentUtils::AnonymizeId(mStableId, aOrigin,
+                              nsContentUtils::OriginFormat::Plain);
+  mStableId.Truncate(kIdLength);
+  return NS_OK;
+}